File: julia_port/ImplicitShellMSD.jl
using BenchmarkTools: time
using Makie
using GeometryBasics
using VectorizedRoutines
using LinearAlgebra
using PrettyTables
import DelimitedFiles: readdlm

# Solution Paramaters
global const tsim = 10              # total sim time
global const dt = 0.1               # timestep
global const p = 1                  # applied pressure
global const fixed = 1:48           # fixed nodes (edges)
global const th = 0.1               # material thickness
global const rho = 10               # material density
global const beta = 0.25            # Newmark integration beta parameter
global const gamma = 0.5            # Newmark integration gamma parameter
global const damp = 1.5             # damping coefficient

function impShell()
    # Import pointslist, connectivitylist, and edgelist from .txt DelimitedFiles
    PL, CL, EL = importMesh() 

    # Calculate spring rest lengths
    L = Array{Float64, 1}(undef,77)  
    L .=  sqrt.(sum(eachrow((PL[:,EL[1,:]]-PL[:,EL[2,:]]).^2)))

    
    # Calculate nodal masses, assemble mass & damping matrices
    m = zeros(Float64, 96, 1)
    for el ∈ 1:size(CL,2)
        ea = PL[:, CL[2,el]] - PL[:, CL[1,el]]
        eb = PL[:, CL[3,el]] - PL[:, CL[1,el]]
        A = norm(×(ea,eb)/2)
        ne = [3*CL[1,el].-[2; 1; 0] 3*CL[2,el].-[2;1;0] 3*CL[3,el].-[2; 1; 0]]
        m[ne] .+= th*rho*A/3 # mass is calculated as thickness*density*area/3 - three is since 3 points on tri --- iterated for each triangle the mass belongs to 
    end
    m[fixed] .= 1e4
    M = Diagonal(vec(m)) # Mass
    
    C = damp*M  # damping

    # 0'd arrays for holding disp/vel
    d = Array{Float64, 1}(undef, length(PL))
    v = Array{Float64, 1}(undef,length(PL))
    
    for t = 0:dt:tsim

        i = reshape(Int[], 0, 2) # index vec for sparse mat
        k = Float64[] # accum vec for sparse mat
        f = zeros(length(PL),1)

        # Assemble stiffness matrix
        # This will 
        for ele = 1:size(EL,2)

            # Store coords of end points
            V = [ PL[:, EL[1,ele]]+d[3*EL[1,ele].-[2; 1; 0;]]   PL[:, EL[2,ele]]+d[3*EL[2,ele].-[2; 1; 0;]] ]

            # Compute stiffness
            ke = SpringStiffnessMatrix(V,L[ele])

            # Nodal degrees of freedom
            ne = [3*EL[1, ele].-[2; 1; 0;]' 3*EL[2,ele].-[2;1;0]']
    
            je = [ne; ne; ne; ne; ne; ne;]

            ie = [ne'; ne'; ne'; ne'; ne'; ne']

            tmp = [ie je[:]]

            i = [i; tmp]
            k = [k; ke[:]]
        end
       
        K = accumarray(i,k)
        display(size(k))
        
    end
    
    # Convert to GeometryBasics.Mesh
    pts = [Point3f0(val[1], val[2], val[3]) for val in eachcol(PL)]
    fcs = [TriangleFace(val[1], val[2], val[3]) for val in eachcol(CL)] 
    msh = GeometryBasics.Mesh(pts, fcs)


    
    # Plot Mesh 
    scn = Makie.mesh(msh, color = 1 : length(msh.position))
    
    Makie.wireframe!(msh)

    display(scn)


end

function importMesh()

    PL = readdlm("points.txt", ',', Float64)'    # Points list
    CL = readdlm("tris.txt", ',', Int8)'         # Connectivity list
    EL = readdlm("edges.txt", ',', Int8)'        # Edge List

    return PL, CL, EL
end

function SpringStiffnessMatrix(V,k)

    L = sqrt(sum((V[:,1] - V[:,2]).^2))

    Cx = (V[1,1] - V[1,2])/L
    Cy = (V[2,1] - V[2,2])/L
    Cz = (V[3,1] - V[3,2])/L

    A = [Cx^2 Cx*Cy Cx*Cz; Cy*Cx Cy^2 Cy*Cz; Cz*Cx Cz*Cy Cz^2]
    
    K = k*[A -A; -A A]

    return K
end

function accumarray(subs, val, sz=(maximum(subs[:,1])))
    A = zeros(eltype(val[1]), sz,sz)
    for i = 1:size(subs,1)
        @inbounds A[i] += val[i]
    end
    @pt A
    return A
end

impShell()

File: julia_port/ImplicitShellMSD.m
%% Implicit Shell MSD Sim
%Using Linear Springs, Newmark Integration, no bending stiffness (yet)
%Lawrence Smith | lasm4254@colorado.edu
%1 July 2021

clear; clc; close all

load mesh.mat
PT = mesh.Points;
CM = mesh.ConnectivityList;

%Set up solution parameters
tsim = 10;                  %total simulation time
dt = 0.1;                   %timestep
p = 1;                      %applied pressure
fixed = 1:48;               %fixed nodes (the edges)
th = 0.1;                   %material thickness
rho = 10;                   %material density
beta = 0.25;                %Newmark Integration Beta Parameter
gamma = 0.5;                %Newmark Integration Gamma Parameter
damp = 1.5;                 %damping coefficient

%Extract edges from mesh
e = edges(mesh);

%Compute initial lengths of each spring
L = sqrt(sum((PT(e(:,1),:)-PT(e(:,2),:))'.^2))';

%Compute the mass of each particle
m = zeros(numel(PT),1);
tic
for el = 1:size(CM,1)                        %index thru triangles
    ea = PT(CM(el,2),:)-PT(CM(el,1),:);         %one edge vector
    eb = PT(CM(el,3),:)-PT(CM(el,1),:);         %another edge vector
    A = norm(cross(ea,eb)/2);                     %area of this triangle
    ne = [3*CM(el,1)-[2 1 0] 3*CM(el,2)-[2 1 0] 3*CM(el,3)-[2 1 0]];
    m(ne) = m(ne) + th*rho*A/3;   %account for this mass
    disp(m(ne))
end
toc


%Adjust the masses of the fixed nodes
m(fixed) = 1e4;


%Form global mass matrix

M = diag(m);

%Form global damping matrix
C = damp*M;

%Initialize a matrix of zeros to hold displacement/velocity solutions
d = zeros(numel(PT),1);
v = zeros(numel(PT),1);

tic
for t = 0:dt:tsim
    
    i = []; %initialize indexing vector for sparse matrix
    k = []; %initialize accumulate vector for sparse matrix
    f = zeros(numel(PT),1);
    
    %Assemble stiffness matrix
    
    for el = 1:size(e,1)
        
        V = [PT(e(el,1),:) + d(3*e(el,1)-[2 1 0])'; 
             PT(e(el,2),:) + d(3*e(el,2)-[2 1 0])';]; %store the coords of the endpoints  
        
        ke = SpringStiffnessMatrix(V,L(el)); %compute stiffness
        
        %identify the nodal degrees of freedom
        ne = [3*e(el,1)-[2 1 0] 3*e(el,2)-[2 1 0]];
        [je, ie] = meshgrid(ne);
        
        %Store the indices and stiffness values into arrays
        i = [i; [ie(:) je(:)]];
        k = [k; ke(:)];
        
    end
    
    %accumulate the stiffness array
%     K = sparse(i(:,1),i(:,2),k); %I wonder if sparse matrix is faster?

    K = accumarray(i,k);

   
    i = []; %initialize indexing vector for sparse matrix
    
    %Assemble the force vector
    
    for el = 1:size(CM,1)
        ea = PT(CM(el,2),:)-PT(CM(el,1),:);
        eb = PT(CM(el,3),:)-PT(CM(el,1),:);
        fe = (p/6)*cross(ea,eb);
        ne = [3*CM(el,1)-[2 1 0] 3*CM(el,2)-[2 1 0] 3*CM(el,3)-[2 1 0]];
        f(ne) = repmat(fe,1,3);
    end
    
    
    if t == 0 %Solve for accelerations at first timestep
        a = inv(M) * (f - (C*v + K*d));
    end
    
    %Solve for Predictors
    v_predict = v + dt * (1-gamma)*a;
    d_predict = d + dt * v + dt^2/2 * (1-2*beta)*a;
    
    
    %Solve for Acceleration
    a = inv(M + gamma*dt*C + beta*dt^2*K) * (f - (C*v_predict + K*d_predict));
    
    %Apply correctors
    v = v_predict + gamma*dt*a;
    d = d_predict + beta*dt^2*a;
      
    %display

    trimesh(triangulation(CM,PT + reshape(d,3,[])'),'facecolor','w','edgecolor','b')
    axis([0 1 0 1 0 1])
    drawnow()


end

toc